<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Hands-on Lab: Interactive Visual Analytics with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [51]:
import piplite
await piplite.install(['folium'])
await piplite.install(['pandas'])

In [52]:
import folium
import pandas as pd

In [53]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/v4/DV0101EN-Exercise-Generating-Maps-in-Python.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [54]:
# Download and read the `spacex_launch_geo.csv`
from js import fetch
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
resp = await fetch(URL)
spacex_csv_file = io.BytesIO((await resp.arrayBuffer()).to_py())
spacex_df=pd.read_csv(spacex_csv_file)

Now, you can take a look at what are the coordinates for each site.


In [55]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [56]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [57]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [58]:
# Initial the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label
# Iterar sobre cada sitio de lanzamiento y agregar un círculo y un marcador
for index, site in launch_sites_df.iterrows():
    # Crear un círculo en la ubicación del sitio de lanzamiento
    folium.Circle(
        location=[site['Lat'], site['Long']],
        radius=1000,
        color='#d35400',
        fill=True
    ).add_child(folium.Popup(site['Launch Site'])).add_to(site_map)
    
    # Crear un marcador con etiqueta de texto para el sitio de lanzamiento
    folium.map.Marker(
        location=[site['Lat'], site['Long']],
        icon=DivIcon(
            icon_size=(20, 20),
            icon_anchor=(0, 0),
            html=f'<div style="font-size: 12; color:#d35400;"><b>{site["Launch Site"]}</b></div>'
        )
    ).add_to(site_map)

# Mostrar el mapa
site_map

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [59]:
spacex_df.tail()

Launch Site        Lat      Long  class
51  CCAFS SLC-40  28.563197 -80.57682      0
52  CCAFS SLC-40  28.563197 -80.57682      0
53  CCAFS SLC-40  28.563197 -80.57682      0
54  CCAFS SLC-40  28.563197 -80.57682      1
55  CCAFS SLC-40  28.563197 -80.57682      0

In [60]:
print(spacex_df)

     Launch Site        Lat        Long  class
0    CCAFS LC-40  28.562302  -80.577356      0
1    CCAFS LC-40  28.562302  -80.577356      0
2    CCAFS LC-40  28.562302  -80.577356      0
3    CCAFS LC-40  28.562302  -80.577356      0
4    CCAFS LC-40  28.562302  -80.577356      0
5    CCAFS LC-40  28.562302  -80.577356      0
6    CCAFS LC-40  28.562302  -80.577356      0
7    CCAFS LC-40  28.562302  -80.577356      0
8    CCAFS LC-40  28.562302  -80.577356      0
9    CCAFS LC-40  28.562302  -80.577356      0
10   CCAFS LC-40  28.562302  -80.577356      0
11   CCAFS LC-40  28.562302  -80.577356      0
12   CCAFS LC-40  28.562302  -80.577356      0
13   CCAFS LC-40  28.562302  -80.577356      0
14   CCAFS LC-40  28.562302  -80.577356      0
15   CCAFS LC-40  28.562302  -80.577356      0
16   CCAFS LC-40  28.562302  -80.577356      0
17   CCAFS LC-40  28.562302  -80.577356      1
18   CCAFS LC-40  28.562302  -80.577356      1
19   CCAFS LC-40  28.562302  -80.577356      0
20   CCAFS LC

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [61]:
marker_cluster = MarkerCluster()


*TODO:* Create a new column in `spacex_df` dataframe called `marker_color` to store the marker colors based on the `class` value


In [62]:

# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red
#----------------------
# Create a new column for marker color based on the 'class' value
spacex_df['marker_color'] = spacex_df['class'].apply(lambda x: 'green' if x == 1 else 'red')

success_rates = spacex_df.groupby('Launch Site')['class'].mean()
print(success_rates)


Launch Site
CCAFS LC-40     0.269231
CCAFS SLC-40    0.428571
KSC LC-39A      0.769231
VAFB SLC-4E     0.400000
Name: class, dtype: float64


*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [63]:
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)

# for each row in spacex_df data frame
for index, record in spacex_df.iterrows():
    # create a Marker object with its coordinate
    marker = folium.Marker(
        location=[record['Lat'], record['Long']],
        popup=f"Launch Site: {record['Launch Site']}<br>Success: {'Yes' if record['class'] == 1 else 'No'}",
        icon=folium.Icon(color='white', icon_color=record['marker_color'], icon='rocket', prefix='fa')
    )
    # Add the marker to the cluster
    marker_cluster.add_child(marker)

site_map


Este código realiza lo siguiente:

Agrega el marker_cluster al site_map.

Itera sobre cada fila del dataframe spacex_df.

Crea un objeto Marker para cada lanzamiento con:

La ubicación del sitio de lanzamiento.

Un popup que muestra el nombre del sitio y si el lanzamiento fue exitoso.

Un icono personalizado: blanco con el color del marcador (verde o rojo) y un símbolo de cohete.

Agrega cada marcador al cluster.

Este enfoque agrupará automáticamente los marcadores cercanos y los expandirá al hacer zoom, mejorando la visualización del mapa

Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png">
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png">
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [64]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


In [65]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


# find coordinate of the closet coastline
# e.g.,: Lat: 28.56367  Lon: -80.57163
# distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)
#------------
# Coordenadas del sitio de lanzamiento (por ejemplo, CCAFS SLC-40)
launch_site_lat = 28.563197
launch_site_lon = -80.576820

# Coordenadas del punto más cercano en la costa (obtenidas usando MousePosition)
coastline_lat = 28.56367
coastline_lon = -80.57163

# Calcular la distancia entre el sitio de lanzamiento y el punto más cercano en la costa
distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)

# Crear y agregar un marcador en el punto más cercano de la costa
distance_marker = folium.Marker(
    [coastline_lat, coastline_lon],
    icon=DivIcon(
        icon_size=(20, 20),
        icon_anchor=(0, 0),
        html=f'<div style="font-size: 12; color:#d35400;"><b>{distance_coastline:.2f} KM</b></div>'
    )
)

# Agregar el marcador al mapa
site_map.add_child(distance_marker)

# Mostrar el mapa actualizado
site_map
#-------------

Explicación del código:
Coordenadas del sitio de lanzamiento:

Utilizamos las coordenadas del sitio de lanzamiento (launch_site_lat y launch_site_lon).

En este caso, se usa CCAFS SLC-40 como ejemplo.

Coordenadas del punto más cercano en la costa:

Estas coordenadas (coastline_lat y coastline_lon) se obtienen utilizando el plugin MousePosition, que muestra las coordenadas al mover el cursor sobre el mapa.

Cálculo de la distancia:

La función calculate_distance calcula la distancia entre las coordenadas del sitio de lanzamiento y las coordenadas del punto más cercano en la costa utilizando la fórmula del Haversine.

Creación del marcador:

Usamos folium.Marker para marcar el punto más cercano en la costa.

El marcador incluye una etiqueta (DivIcon) que muestra la distancia calculada en kilómetros.

Agregar el marcador al mapa:

El marcador se agrega al mapa usando site_map.add_child(distance_marker).

Visualización:

El mapa actualizado muestra el marcador con la distancia al punto más cercano en la costa.

Este enfoque te permitirá visualizar claramente la proximidad entre los sitios de lanzamiento y los puntos cercanos en la costa, lo cual es útil para analizar ubicaciones estratégicas.

In [66]:
launch_sites = {
    "W_LC": {
        "launch_loc": [28.56234, -80.57738],
        "destinations": {
            "coast": [28.56276, -80.56785],
            "rail": [28.57116, -80.58538],
            "highway": [28.56327, -80.57084],
            "city": [28.4054, -80.59317]
        }
    },
    "W_SLC": {
        "launch_loc": [28.56321, -80.57682],
        "destinations": {
            "coast": [28.56276, -80.56785],
            "rail": [28.57116, -80.58538],
            "highway": [28.56327, -80.57084],
            "city": [28.4054, -80.59317]
        }
    },
    "W_KSC": {
        "launch_loc": [28.57329, -80.64692],
        "destinations": {
            "coast": [28.57144, -80.56927],
            "rail": [28.58046, -80.59124],
            "highway": [28.57295, -80.65540],
            "city": [28.61202, -80.80573]
        }
    },
    "E_VAFB": {
        "launch_loc": [34.63286, -120.61072],
        "destinations": {
            "coast": [34.63639, -120.62508],
            "rail": [34.63642, -120.62370],
            "highway": [34.62812, -120.43367],
            "city": [34.63225, -120.48363]
        }
    }
}

In [67]:
# Coordenadas del sitio de lanzamiento CCAFS SLC-40
launch_site_lat, launch_site_lon = launch_sites["W_SLC"]["launch_loc"]

# Coordenadas del punto más cercano en la costa
coastline_lat, coastline_lon = launch_sites["W_SLC"]["destinations"]["coast"]

# Calcular la distancia
distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)

print(f"La distancia entre CCAFS SLC-40 y el punto más cercano en la costa es aproximadamente {distance_coastline:.2f} km.")

# Crear y agregar un marcador en el punto más cercano de la costa
distance_marker = folium.Marker(
    [coastline_lat, coastline_lon],
    icon=DivIcon(
        icon_size=(20, 20),
        icon_anchor=(0, 0),
        html=f'<div style="font-size: 12; color:#d35400;"><b>{distance_coastline:.2f} KM</b></div>'
    )
)

# Agregar el marcador al mapa
site_map.add_child(distance_marker)

# Crear una línea entre el sitio de lanzamiento y el punto en la costa
line = folium.PolyLine(
    locations=[[launch_site_lat, launch_site_lon], [coastline_lat, coastline_lon]],
    weight=2,
    color='red',
    opacity=0.8,
    tooltip='Distancia al punto más cercano de la costa'
)

# Añadir la línea al mapa
site_map.add_child(line)

# Mostrar el mapa actualizado
site_map


La distancia entre CCAFS SLC-40 y el punto más cercano en la costa es aproximadamente 0.88 km.


Este código realiza las siguientes acciones:

Extrae las coordenadas del sitio de lanzamiento y del punto más cercano en la costa del diccionario proporcionado.

Calcula la distancia entre estos dos puntos usando la función calculate_distance.

Crea un marcador en el punto de la costa con la distancia calculada.

Dibuja una línea entre el sitio de lanzamiento y el punto en la costa.

Agrega tanto el marcador como la línea al mapa.

Este enfoque te permitirá visualizar claramente la distancia entre CCAFS SLC-40 y el punto más cercano en la costa, así como su ubicación relativa en el mapa.

In [ ]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
# distance_marker = folium.Marker(
#    coordinate,
#    icon=DivIcon(
#        icon_size=(20,20),
#        icon_anchor=(0,0),
#        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance),
#        )
#    )

*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
# lines=folium.PolyLine(locations=coordinates, weight=1)
#site_map.add_child(lines)
#--------------------
# Coordenadas del sitio de lanzamiento (CCAFS SLC-40)
launch_site_lat = 28.563197
launch_site_lon = -80.576820

# Coordenadas del punto más cercano en la costa (obtenidas anteriormente)
coastline_lat = 28.56367
coastline_lon = -80.57163

# Crear una lista de coordenadas para la línea
coordinates = [
    [launch_site_lat, launch_site_lon],
    [coastline_lat, coastline_lon]
]

# Crear un objeto PolyLine
line = folium.PolyLine(
    locations=coordinates,
    weight=2,
    color='red',
    opacity=0.8,
    tooltip='Distancia al punto más cercano de la costa'
)

# Añadir la línea al mapa
site_map.add_child(line)

# Mostrar el mapa actualizado
site_map


Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png">
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png">
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png">
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png">
</center>


In [48]:
# Iterar sobre todos los sitios de lanzamiento
for site_name, site_data in launch_sites.items():
    launch_site_lat, launch_site_lon = site_data["launch_loc"]
    
    # Crear marcador para el sitio de lanzamiento
    folium.Marker(
        [launch_site_lat, launch_site_lon],
        popup=f"Sitio de lanzamiento: {site_name}",
        icon=folium.Icon(color='red', icon='rocket', prefix='fa')
    ).add_to(site_map)

    # Puntos de interés para este sitio
    points_of_interest = site_data["destinations"]

    for poi_name, poi_coords in points_of_interest.items():
        poi_lat, poi_lon = poi_coords
        distance = calculate_distance(launch_site_lat, launch_site_lon, poi_lat, poi_lon)
        
        # Crear marcador para el punto de interés con la distancia calculada
        folium.Marker(
            [poi_lat, poi_lon],
            popup=f"{site_name} - {poi_name.capitalize()}: {distance:.2f} km",
            icon=folium.DivIcon(
                icon_size=(150,36),
                icon_anchor=(0,0),
                html=f'<div style="font-size: 12px; color:#d35400;"><b>{poi_name.capitalize()}: {distance:.2f} km</b></div>'
            )
        ).add_to(site_map)
        
        # Dibujar línea entre el sitio de lanzamiento y el punto de interés
        folium.PolyLine(
            locations=[[launch_site_lat, launch_site_lon], [poi_lat, poi_lon]],
            color='blue',
            weight=2,
            opacity=0.8,
            tooltip=f"{site_name} - Distancia a {poi_name}: {distance:.2f} km"
        ).add_to(site_map)

# Mostrar el mapa actualizado
site_map



In [68]:
# Coordenadas del sitio de lanzamiento CCAFS SLC-40
launch_site_lat, launch_site_lon = launch_sites["W_SLC"]["launch_loc"]

# Crear marcador para el sitio de lanzamiento
folium.Marker(
    [launch_site_lat, launch_site_lon],
    popup="Sitio de lanzamiento: CCAFS SLC-40",
    icon=folium.Icon(color='red', icon='rocket', prefix='fa')
).add_to(site_map)

# Puntos de interés para CCAFS SLC-40
points_of_interest = launch_sites["W_SLC"]["destinations"]

for poi_name, poi_coords in points_of_interest.items():
    poi_lat, poi_lon = poi_coords
    distance = calculate_distance(launch_site_lat, launch_site_lon, poi_lat, poi_lon)
    
    # Crear marcador para el punto de interés con la distancia calculada
    folium.Marker(
        [poi_lat, poi_lon],
        popup=f"{poi_name.capitalize()}: {distance:.2f} km",
        icon=folium.DivIcon(
            icon_size=(150,36),
            icon_anchor=(0,0),
            html=f'<div style="font-size: 12px; color:#d35400;"><b>{poi_name.capitalize()}: {distance:.2f} km</b></div>'
        )
    ).add_to(site_map)
    
    # Dibujar línea entre el sitio de lanzamiento y el punto de interés
    folium.PolyLine(
        locations=[[launch_site_lat, launch_site_lon], [poi_lat, poi_lon]],
        color='blue',
        weight=2,
        opacity=0.8,
        tooltip=f"Distancia a {poi_name}: {distance:.2f} km"
    ).add_to(site_map)

# Mostrar el mapa actualizado
site_map



After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/)


<!--## Change Log--!>


<!--| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|--!>


### <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
